Add a usage and advice notice to each module

thatux opened the pull request:
Added optional possibility to add additional info this allows module authors to add usage instructions/how to generate hash to their module.

Furthermore, it allows us to include the footnotes from https://hashcat.net/wiki/doku.php?id=example_hashes in to the modules themselves such that they do not need to be manually administered by @roycewilliams (https://github.com/hashcat/hashcat/issues/4467).

Example passwords and hashes are already maintained within each module, so makes sense to also keep some instructions there; it is very similar to `deprecated_notice`.

This implementation adds module_comment as an optional callback. Not every module needs to declare it in it's source: defaults of `N/A`. The only thing I think this breaks are pre-compiled modules (without any .c source), but I think those are very rare.

Example of footnotes:
```
4 The hash used here is not the one sent via e.g. the web interface to LastPass servers (pbkdf2_sha256_hex (pbkdf2_sha256 ($pass, $email, $iterations), $pass, 1) but instead the one stored (by e.g. your browser or the pocket version) to disk. For instance, Opera and Chrome store the hash in local SQLite databases; Firefox uses files ending with “lpall.slps” - for Linux: 2nd line is interesting / base64 decode it; for Windows, see here - and_key.itr
5 You can consider the second part as a “salt”. If it is equal to 00000000, the CRC32 code will be considered as “not salted”
6 The raw sha256 output is used for base64() encoding (not the hexadecimal output)
7 The format is hash:salt:id
10 This file actually contains several examples of the different hash+cipher combinations. The password is stored in the pw file.
11 You can use itunes_backup2hashcat to extract the hashes from the Manifest.plist file
12 Password: “hashcat!!!”. Min/max password length is exactly 10 characters/bytes.
13 You can use AxSuite by Fist0urs to retrieve the hashes.
```
Example output
```
$ ./hashcat -HH -m34300 | grep -i usage                                                          
  Usage.Notice........: To extract hash from KDBX use https://github.com/openwall/john/blob/bleeding-jumbo/src/keepass2john.c
```

```
$ ./hashcat -a3 -m1000 'b4b9b02e6f09a9bd760f388b67351e2b' '?a?a?a?a?a?a?a'                       
hashcat (v7.0.0-beta.1-306-g972f67ba0+) starting

[snip]

Module 1000 advice notice: When using the optimized-kernel: some byte encoding is simplified that would make some non-ASCII passwords never be cracked, even if they are under the optimized byte limit. The pure-kernel implements true UTF16 handling so doesn't have this problem and should crack all passwords.
```


<details>

TODO
- [x] Might be good to limit the character limit on the comment, to 140 characters. I'm not sure whether we really want this, maybe manual check per commit is fine. If we want to do it automatically we could use a linter? --> too preliminar to implement, an URL like this is already 71 characters 'https://github.com/openwall/john/blob/bleeding-jumbo/src/keepass2john.c'

</details>

Commit:
add module comment functionality

PenguinKeeper7 commented:
Related: https://github.com/hashcat/hashcat/issues/3958

roycewilliams commented:
> Related: #3958

Good callback! Maybe two different fields - general note, vs caveats/warnings?

thatux commented:
I don't have a very strong opinion, but I'm a bit hesitant to split between warning and a  note: not sure how much both will be used. Is it worthwhile to make that distinction?

Are there many more modules where we specifically like a warning to be printed? @roycewilliams 

I think most modules would at least benefit a note describing how to generate the hash i.e. what hashcat2x or john2x script to use.



Implementation could look like this:
- Warnings are always printed when running a attack.
- Usage notes are printed upon request -HH -m, or when the attack is run without a hash

I like usage_note better than comment so I'll change that.

Commit:
rename module_usage_notice to module_notice_usage

Commit:
add Notice.Warning

roycewilliams commented:
> Yeah, that makes sense!

thatux commented:
Maybe call the warning a 'module advice' instead? I now use event_log_advice() instead of event_log_warning(), which makes sense to me

Commit:
renamed notice.warning to advice.notice

Commit:
added footnotes from https://hashcat.net/wiki/doku.php\?id\=example_hashes to the specific modules

thatux commented:
@roycewilliams I've added all footnotes from the table at https://hashcat.net/wiki/doku.php?id=example_hashes to the relevant files; I don't think these are currently in use:
```
<sup>3</sup> Same format as in <sup>2</sup> but the number of rounds <strong>must</strong> be specified<br/>
<sup>6</sup> The raw sha256 output is used for base64() encoding (not the hexadecimal output)<br/>
<sup>21</sup> you can extract the hashes with <a href="[https://github.com/0x6470/bitwarden2hashcat](view-source:https://github.com/0x6470/bitwarden2hashcat)" class="urlextern" title="https://github.com/0x6470/bitwarden2hashcat" rel="ugc nofollow">https://github.com/0x6470/bitwarden2hashcat</a><br/>
```

I've added `<sup>21</sup>` to `23400`, the other two I'm not sure about..

Commit:
add usage.notice and advice.notice to docs

Commit:
Merge branch 'master' into add_comment_field

Commit:
docs/hashcat-example-hashes.md with auto-generated footnotes

thatux commented:
Looks good I think: https://github.com/hashcat/hashcat/blob/401acd32be2d3fa63adcdf81c3424ac3c4b0f2fe/docs/hashcat-example-hashes.md

Commit:
remove backticks and pipes from developer-input; it would break the markdown table

jsteube commented on include/common.h:
```diff
@@ -145,3 +145,6 @@ but this is needed for VS compiler which doesn't have inline keyword but has __i
 #endif // HC_COMMON_H
 
+
+#define USAGE_NOTICE_NUM_ROUNDS "rounds=[# of iterations] is optional here, after signature, e.g. $5$rounds=5000"
```
Cool example!

jsteube commented on src/interface.c:
```diff
@@ -122,6 +122,8 @@ int hashconfig_init (hashcat_ctx_t *hashcat_ctx)
 
   if (rc_load == false) return -1;
 
+  module_ctx->module_usage_notice = MODULE_DEFAULT; // set all module to have usage_notice by empty default; such that this property doesn't have to be declared explicitly by all modules (such that we don't break private plugins without this options)
```
I kind of like the idea that we no longer need to update all existing plugins because of a newly added callback, however it's the first time we would do this. Did you check if this works on windows, too?

thatux replied:
Thanks! Just tested on Windows and there it also works  (built for Windows with [WSL](https://github.com/hashcat/hashcat/blob/master/BUILD_WSL.md):
```
$ file hashcat.exe
hashcat.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows

$ ./hashcat.exe -HH -m34300 | grep -i usage
  Usage.Notice........: To extract hash from KDBX use https://github.com/openwall/john/blob/bleeding-jumbo/src/keepass2john.c

$ ./hashcat.exe -a3 -m1000 'b4b9b02e6f09a9bd760f388b67351e2b' '?a?a?a?a?a?a?a'
Module advice notice: Some byte encoding is simplified in the optimized kernel for NTLM that would make some non-ASCII passwords never be cracked, even if they are under the optimized byte limit.
```

jsteube commented on src/modules/module_01000.c:
```diff
@@ -52,6 +52,12 @@ u32         module_salt_type      (MAYBE_UNUSED const hashconfig_t *hashconfig,
 
+const char *module_advice_notice (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra)
+{
+  return "Some byte encoding is simplified in the optimized kernel for NTLM that would make some non-ASCII passwords never be cracked, even if they are under the optimized byte limit.";
```
Only in -O mode. Pure mode implements true UTF16 handling.

thatux replied:
Thanks, tried to make this more clear by starting with this distinction.

jsteube commented on src/modules/module_14800.c:
```diff
@@ -49,6 +49,11 @@ u32         module_salt_type      (MAYBE_UNUSED const hashconfig_t *hashconfig,
 
+const char *module_usage_notice (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra)
+{
+  return "You can use https://github.com/philsmd/itunes_backup2hashcat/ to extract the hashes from the Manifest.plist file";
```
Phil also has https://github.com/philsmd/7z2hashcat

thatux replied:
Thanks! I've added that to 11600. I'm sure there are many more examples :smile: let me know!

Commit:
Resolve comments by Atom

Commit:
Improve module advice print to include module-number

Commit:
Corrected example-hashes to include advice notice

Commit:
add usage notice for apfs, bitlocker, msonline

jsteube commented:
OK, let's merge and see if there's anything we didn't think of..
